contracts/randdrop/src/contract.rs

```diff
@@ -0,0 +1,763 @@
+use cosmwasm_std::{
+    ensure_eq, entry_point, to_binary, Addr, Attribute, BankMsg, Coin, Deps, DepsMut, Env,
+    HexBinary, MessageInfo, QueryResponse, Response, StdResult, Uint128, WasmMsg,
+};
+use nois::{NoisCallback, ProxyExecuteMsg};
+use sha2::{Digest, Sha256};
+
+use crate::error::ContractError;
+use crate::msg::{
+    ConfigResponse, ExecuteMsg, InstantiateMsg, IsClaimedResponse, IsWinnerResponse,
+    MerkleRootResponse, QueryMsg,
+};
+use crate::state::{
+    Config, NoisProxy, ParticipantData, CLAIMED, CONFIG, MERKLE_ROOT, PARTICIPANTS,
+};
+
+/// The winning chance is 1/AIRDROP_ODDS
+const AIRDROP_ODDS: u64 = 3;
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let manager = deps.api.addr_validate(&msg.manager)?;
+    let randdrop_denom = msg.randdrop_denom;
+    let nois_proxy_address = deps
+        .api
+        .addr_validate(&msg.nois_proxy_address)
+        .map_err(|_| ContractError::InvalidProxyAddress)?;
+    let nois_proxy_price = Coin {
+        denom: msg.nois_proxy_denom,
+        amount: msg.nois_proxy_amount,
+    };
+    let nois_proxy = NoisProxy {
+        address: nois_proxy_address,
+        price: nois_proxy_price,
+    };
+
+    let config = Config {
+        manager,
+        randdrop_denom,
+        nois_proxy,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::default())
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig {
+            manager,
+            nois_proxy_denom,
+            nois_proxy_amount,
+            nois_proxy_address,
+            randdrop_denom,
+        } => execute_update_config(
+            deps,
+            env,
+            info,
+            manager,
+            randdrop_denom,
+            nois_proxy_denom,
+            nois_proxy_amount,
+            nois_proxy_address,
+        ),
+        ExecuteMsg::RegisterMerkleRoot { merkle_root } => {
+            execute_register_merkle_root(deps, env, info, merkle_root)
+        }
+        //RandDrop should be called by the manager with a future timestamp
+        ExecuteMsg::Randdrop { amount, proof } => execute_randdrop(deps, info, amount, proof),
+        //NoisReceive should be called by the proxy contract. The proxy is forwarding the randomness from the nois chain to this contract.
+        ExecuteMsg::NoisReceive { callback } => execute_receive(deps, env, info, callback),
+        ExecuteMsg::Claim {} => execute_claim(deps, env, info),
```

Comment:
Would be cool if this was not needed, i.e. the callback performs the send

Reply:
Oh yeah when I thought about this I still had the merkle check on the claim part, and  I thought it would be hard for the relayer to fetch the json. But indeed now that the merkle is verified before  we can allow the relayer to send the fund. I just hope this doesnt over simplify the process  for the user such that they don't understand the process. But in any case it makes sense to make the user execute only 1 tx.

Reply:
Yeah, so the merkle proof check checks eligibility, i.e. the permission to start the process. This should be done in the first transaction. Then once we get the randomness we know the user is allowed.

Reply:
yep